clang/test/SemaSYCL/nested-anon-and-std-ns.cpp

```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -verify %s
+
+#include "sycl.hpp"
+
+namespace std {
+namespace NestedInStd {
+struct NestedStruct {};
+}; // namespace NestedInStd
+}; // namespace std
+
+namespace {
+namespace NestedInAnon {
+struct StructInAnonymousNS {};
+} // namespace NestedInAnon
+
+} // namespace
+
+namespace ValidNS {
+struct StructinValidNS {};
+} // namespace ValidNS
+
+struct MyWrapper {
+
+public:
+  void test() {
+    cl::sycl::queue q;
+
+    // expected-error@Inputs/sycl.hpp:220 {{'std::NestedInStd::NestedStruct' is an invalid kernel name type}}
+    // expected-note@Inputs/sycl.hpp:220 {{'std::NestedInStd::NestedStruct' cannot be in the "std" namespace}}
+    // expected-note@+2{{in instantiation of function template specialization}}
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<std::NestedInStd::NestedStruct>([] {});
+    });
+
+    // expected-error@Inputs/sycl.hpp:220 {{'(anonymous namespace)::NestedInAnon::StructInAnonymousNS' is an invalid kernel name type}}
+    // expected-note@Inputs/sycl.hpp:220 {{'(anonymous namespace)::NestedInAnon::StructInAnonymousNS' should be globally-visible}}
+    // expected-note@+2{{in instantiation of function template specialization}}
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<NestedInAnon::StructInAnonymousNS>([] {});
+    });
+
+    // no error for valid ns
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<ValidNS::StructinValidNS>([] {});
+    });
+  }
+};
+
+int main() {
+  cl::sycl::queue q;
+
+  return 0;
+}
```

Comment:
Please add a newline to the end of the file.

Reply:
Just curious, I have followed this rule before, but have never been told why. 
clang-format doesn't complain about this, so is it a general coding practise?

Reply:
Yeah, it's one of those weird historical quirks -- old utilities that operate on text files may ignore text after the last newline due to the way UNIX worked, basically (text files are a series of lines that end with a newline, so if the file has a line without a terminating newline, some utilities would not handle it properly). So C and C++ used to both require that any source file be terminated by a newline character, but these days, it's only strictly required by C (5.1.1.2p2) because C++ worked around it by requiring implementations to do some extra work (http://eel.is/c++draft/lex.phases#1.2.sentence-4).

So these days, it's often not strictly necessary, but tools still warn about missing the newline, so it's best to just add them to the end of all your source files.